MennaMohie/assignment2/Solution.cpp

```diff
@@ -0,0 +1,174 @@
+
+public:
+    Node():left(nullptr),right(nullptr) {}
+    Node(T val):left(nullptr),right(nullptr),value(val) {}
+
+    T getValue(){return value;}
+};
+
+template<class T>
+class BinaryTree
+{
+    Node<T> *root;
+
+public:
+    BinaryTree(T val):root(new Node<T>(val)) {}
+
+    Node<T>* getRoot()
+    {
+        return root;
+    }
+
+    bool contains(Node<T> *currentRoot, T val)
+    {
+        if(currentRoot == nullptr)
+            return false;
+        if(currentRoot->value == val)
+            return true;
+        return (contains(currentRoot->left, val) || contains(currentRoot->right, val));
+    }
+
+    bool resetRoot(T val)
+    {
+        if(root==nullptr)
+        {
+            root=new Node<T>(val);
+            return true;
+        }
+        else if(!contains(root, val))
+        {
+            root->value=val;
+            return true;
+        }
+        else
+            return false;
+    }
+
+    //adds a new node (val) as a left child to an existing node (parentVal)
+    bool addAsLeftChild(Node<T> *currentRoot, T parentVal, T val)
+    {
+        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
+            return false;
+        if(currentRoot->value==parentVal)
+        {
+            if(currentRoot->left==nullptr)
+            {
+                currentRoot->left=new Node<T>(val);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        if(addAsLeftChild(currentRoot->left, parentVal, val))
+            return true;
+        return addAsLeftChild(currentRoot->right, parentVal, val);
+
+    }
+
+    bool addAsRightChild(Node<T> *currentRoot, T parentVal, T val)
+    {
+        if(!contains(root, parentVal) || contains(root, val) || currentRoot==nullptr)
+            return false;
+        if(currentRoot->value==parentVal)
+        {
+            if(currentRoot->right==nullptr)
+            {
+                currentRoot->right=new Node<T>(val);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        if(addAsRightChild(currentRoot->left, parentVal, val))
+            return true;
+        else
+            return addAsRightChild(currentRoot->right, parentVal, val);
+
+    }
+
+    void traverse(Node<T> *currentRoot)
+    {
+        if(currentRoot==nullptr)
+            return;
+        cout<<currentRoot->value<<endl;
+        traverse(currentRoot->left);
+        traverse(currentRoot->right);
+    }
+
+    bool printAncestors(Node<T> *currentRoot, T val)
+    {
+        if(currentRoot==nullptr)
+            return false;
+
+        if(currentRoot->value==val)
+            return true;
+
+        bool foundInLeft = printAncestors(currentRoot->left, val);
+        bool foundInRight = printAncestors(currentRoot->right, val);
+        if(foundInLeft || foundInRight)
+        {
+            cout<<currentRoot->value<<endl;
```

Comment:
Please return the path from the printAncestors function, and print it on main. This way your function can be reused.
Since you are already using the return for success/failure, the easies way to do this is to have a `vector<T> *` out parameter.